apis/python/src/tiledbsoma/io/ingest.py

```diff
@@ -1681,6 +1681,70 @@ def update_matrix(
     )
 
 
+def _validate_matrix_to_collection(
+    exp: Experiment,
+    measurement_name: str,
+    collection_name: str,
+    matrix_name: str,
+    matrix_data: Matrix | h5py.Dataset,
+) -> None:
+    """Validates the shape of a new matrix against an existing matrix in a collection.
+
+    Raises:
+        ValueError: If the matrix shape is incompatible.
+    """
+    obs_soma_joinid_domain = exp.obs._maybe_soma_joinid_shape  # O
+    var_soma_joinid_domain = exp.ms[measurement_name].var._maybe_soma_joinid_shape  # V
+
+    target_obs_size = obs_soma_joinid_domain  # O: Observation (row) size
+    target_var_size = var_soma_joinid_domain  # V: Variable (column) size
+    target_shape = (target_obs_size, target_var_size)
+    # Check if the dimensionality is the same
+    if len(matrix_data.shape) != len(target_shape):
+        raise SOMAError(
+            f"Matrix '{matrix_name}' has {len(matrix_data.shape)} dimensions, "
+            f"but the existing matrix has {len(target_shape)}. Shapes must match."
+        )
+
+    if collection_name == "X":
+        # X: (O, V) - Must match both dimensions exactly
+        if matrix_data.shape != target_shape:
+            raise SOMAError(
+                f"Matrix 'X' shape {matrix_data.shape} is incompatible with target shape {target_shape}. "
+                "Both dimensions must match exactly."
+            )
+    elif collection_name == "obsm":
+        if matrix_data.shape[0] != target_obs_size:
+            raise SOMAError(
+                f"Matrix '{matrix_name}' must match the observation size (O). "
+                f"Found {matrix_data.shape[0]}, expected {target_obs_size}."
+            )
+    elif collection_name == "varm":
+        if matrix_data.shape[0] != target_var_size:
+            raise SOMAError(
+                f"Matrix '{matrix_name}' must match the variable size (V). "
+                f"Found {matrix_data.shape[0]}, expected {target_var_size}."
+            )
+    elif collection_name == "obsp":
```

Comment:
`obsp` and `varp` checks are incorrect.  These arrays can have ndim>2.  The check should be that the first two dimensions match the respective axis shape.

I.e., for `obsp`, you want to check that `matrix_data.shape[0:2] == (target_obs_size, target_obs_size)`  (note the slice on the matrix shape).